cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -207,6 +208,183 @@ func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {
 		}
 	}
 
+	// If leader migration is enabled, use the redirected initialization
+	// otherwise, make absolutely NO change to the initialization process
+	if leadermigration.FeatureEnabled() && c.ComponentConfig.Generic.LeaderMigrationEnabled &&
+		c.ComponentConfig.Generic.LeaderElection.LeaderElect {
+
+		klog.Infof("starting leader migration")
+		leaderMigrator := leadermigration.NewLeaderMigrator(&c.ComponentConfig.Generic.LeaderMigration,
+			"kube-controller-manager")
+
+		// We need a channel to signal the start of Service Account Token Controller
+		//  all migrated channel must start after this.
+		saTokenControllerStarted := make(chan struct{})
+
+		// Extracted helper, we do not want to touch the path without leader migration
+		//  so the helper will only be used in run and migrated run.
+		createControllerContext := func(ctx context.Context, rootClientBuilder clientbuilder.ControllerClientBuilder) ControllerContext {
+			var clientBuilder clientbuilder.ControllerClientBuilder
+			if c.ComponentConfig.KubeCloudShared.UseServiceAccountCredentials {
+				if len(c.ComponentConfig.SAController.ServiceAccountKeyFile) == 0 {
+					// It's possible another controller process is creating the tokens for us.
+					// If one isn't, we'll timeout and exit when our client builder is unable to create the tokens.
+					klog.Warningf("--use-service-account-credentials was specified without providing a --service-account-private-key-file")
+				}
+
+				clientBuilder = clientbuilder.NewDynamicClientBuilder(
+					restclient.AnonymousClientConfig(c.Kubeconfig),
+					c.Client.CoreV1(),
+					metav1.NamespaceSystem)
+			} else {
+				clientBuilder = rootClientBuilder
+			}
+			controllerContext, err := CreateControllerContext(c, rootClientBuilder, clientBuilder, ctx.Done())
+			if err != nil {
+				klog.Fatalf("error building controller context: %v", err)
+			}
+			return controllerContext
+		}
+
+		// copied from the original func run, with extracted createControllerContext
+		run := func(ctx context.Context) {
```

Comment:
Having two different run variables in the same methods seems confusing. This should probably be migrationRun.

Reply:
I agree. This can be resolved by extracting the block into a separate function.